File: examples/sendrecv_global_cells/python/cfd_sendrecv_cells.py
from mpi4py import MPI
from cplpy.cpl import CPL 
import numpy as np

CPL = CPL()
nsteps = 1
dt = 0.2
density = 0.8

# Parameters of the cpu topology (cartesian grid)
NPx = 2
NPy = 2
NPz = 1
NProcs = NPx*NPy*NPz

# Parameters of the mesh topology (cartesian grid)
ncxyz = np.array([64, 18, 64], order='F', dtype=np.int32)
xyzL = np.array([10.0, 10.0, 10.0], order='F', dtype=np.float64)
xyz_orig = np.array([0.0, 0.0, 0.0], order='F', dtype=np.float64)

# Create communicators and check that number of processors is consistent
realm_comm = CPL.init(CPL.CFD_REALM)
nprocs_realm = realm_comm.Get_size()

if (nprocs_realm != NProcs):
    print "ERROR: Non-coherent number of processors."
    MPI.Abort(errorcode=1)

cart_comm = realm_comm.Create_cart([NPx, NPy, NPz])
CPL.setup_cfd(nsteps, dt, cart_comm, xyzL, xyz_orig, ncxyz, density)

cart_rank = cart_comm.Get_rank()
limits = CPL.get_olap_limits()
portion = CPL.my_proc_portion(limits)
[ncxl, ncyl, nczl] = CPL.get_no_cells(portion)
send_array = np.ones((3, ncxl, ncyl, nczl), order='F', dtype=np.float64)

for i in range(portion[0],portion[1]+1):
    for j in range(portion[2],portion[3]+1):
        for k in range(portion[4],portion[5]+1):
            ii = i - portion[0]
            jj = j - portion[2]
            kk = k - portion[4]

            send_array[0,ii,jj,kk] = i+1
            send_array[1,ii,jj,kk] = j+1
            send_array[2,ii,jj,kk] = k+1

ierr = CPL.send(send_array, limits)

MPI.COMM_WORLD.Barrier()

File: test/valgrind/cfd_sendrecv_cells.f90
program cfd_cpl_example
    use cpl, only : CPL_init, CPL_setup_cfd, & 
                    CPL_get_olap_limits, CPL_my_proc_portion, &
                    CPL_get_no_cells, CPL_send, CPL_recv, &
					CPL_finalize, CPL_overlap
    use array_stuff
    use mpi
    implicit none

    logical :: recv_flag, send_flag, NO_ERROR
    integer :: i,j,k,ii,jj,kk,ierr
    integer :: NPx, NPy, NPz, NProcs, rank
    integer :: nprocs_realm, nsteps, initialstep
    integer :: CART_COMM, CFD_COMM
    integer, parameter :: cfd_realm=1
    integer, dimension(3) :: npxyz, Ncells, ncxyz
    integer, dimension(6) :: portion, limits
    double precision :: dt, density
    double precision, dimension(3)  :: xyzL, xyz_orig
    double precision, dimension(:,:,:,:), & 
        allocatable  :: send_array, recv_array

    !Initialise MPI
    call MPI_Init(ierr)

    !Create MD Comm by spliting world
    call CPL_init(CFD_realm, CFD_COMM, ierr)

    !Parameters
    dt = 0.1
    density = 0.8
    nsteps = 100

    ! Parameters of the cpu topology (cartesian grid)
    call read_input(xyzL=xyzL, xyz_orig=xyz_orig, & 
                    npxyz_CFD=npxyz, ncxyz=ncxyz)

    ! Create communicators and check that number of processors is consistent
    call MPI_Comm_size(CFD_COMM, nprocs_realm, ierr) 
    if (nprocs_realm .ne. product(npxyz)) then
        print'(4(a,i6))', "Non-coherent number of processes in CFD ", nprocs_realm, & 
                " no equal to ",  npxyz(1), " X ", npxyz(2), " X ", npxyz(3)
        call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
    endif

    !Setup cartesian topology
    call MPI_comm_rank(CFD_COMM, rank, ierr)
    call MPI_Cart_create(CFD_COMM, 3, npxyz, (/1, 1, 1/), & 
                         .true., CART_COMM, ierr)

    !Coupler setup
    call CPL_setup_cfd(nsteps, dt, CART_COMM, xyzL, xyz_orig, ncxyz, density)

    !Get detail for grid
    call CPL_get_olap_limits(limits)
    call CPL_my_proc_portion(limits, portion)
    call CPL_get_no_cells(portion, Ncells)

    !Pack array with cell data
    call fill_array(send_array)
    call CPL_send(send_array, limits, send_flag)

    !Block before checking if successful
    call MPI_Barrier(MPI_COMM_WORLD, ierr)

    !Coupled Recieve and print
    allocate(recv_array(3, Ncells(1), Ncells(2), Ncells(3)))
    recv_array = 0.d0
    call CPL_recv(recv_array, limits, recv_flag)
    call print_array(recv_array, rank, no_error)

    !Block before checking if successful
    call MPI_Barrier(MPI_COMM_WORLD, ierr)
    if (CPL_overlap() .and. no_error) then
        print'(a,a,i2,a)', "CFD -- ", "(rank=", rank, ") CELLS HAVE BEEN RECEIVED CORRECTLY."
    endif

    !Release all coupler comms 
    call CPL_finalize(ierr)

    !Deallocate arrays and finalise MPI
    deallocate(send_array, recv_array)
	call MPI_COMM_FREE(CART_COMM, ierr)
	call MPI_COMM_FREE(CFD_COMM, ierr)
    call MPI_finalize(ierr)

end program cfd_cpl_example
